carbonserver/carbonserver.go

```diff
@@ -591,19 +607,51 @@ func (listener *CarbonserverListener) UpdateMetricsAccessTimesByRequest(metrics
 	listener.UpdateMetricsAccessTimes(accessTimes, false)
 }
 
+func splitAndInsert(cacheMetricNames map[string]struct{}, newCacheMetricNames []map[string]struct{}) map[string]struct{} {
+	// splits each new metric from cache-scan and inserts
+	// into the current cacheMetricNames map
+	// in: new.metric.name1 --> split by "."
+	// insert "/new" , "/new/metric", "/new/metric/name1.wsp" into the
+	// metricsName map. This is inline with the inserts
+	// during filescan walk
+	for _, shardAddMap := range newCacheMetricNames {
+		for newMetric := range shardAddMap{
+			split := strings.Split(newMetric, ".")
+			fileName := "/"
+			for i, seg := range split {
+				fileName = filepath.Join(fileName, seg)
+				if i == len(split) - 1 {
+					fileName = fileName + ".wsp"
+				}
+				if _, ok := cacheMetricNames[fileName]; !ok {
+					cacheMetricNames[fileName] = struct{}{}
+				}
+			}
+		}
+	}
+	return cacheMetricNames
+}
+
 func (listener *CarbonserverListener) fileListUpdater(dir string, tick <-chan time.Time, force <-chan struct{}, exit <-chan struct{}) {
+	cacheMetrics := make(map[string]struct{})
 	for {
 		select {
 		case <-exit:
 			return
 		case <-tick:
 		case <-force:
 		}
-		listener.updateFileList(dir)
+		if listener.cacheGetRecentMetrics != nil{
+			// cacheMetrics maintains all new metric names added in cache
+			// when cache-scan is enabled in conf
+			newCacheMetricNames := listener.cacheGetRecentMetrics()
+			cacheMetrics = splitAndInsert(cacheMetrics, newCacheMetricNames)
```

Comment:
nits: why not just calling this func in updateFileList, this way you don't have to change func signatures of listener.updateFileList.

Reply:
So overtime cacheMetrics would contain all the files and directories of the whole whisper data directory? Probably few hundreds megabytes or more of memory overhead for 4-5 millions of metrics?

Reply:
That is correct! There is memory overhead involved.